planner/core/indexmerge_path.go

```diff
@@ -467,3 +488,145 @@ func (ds *DataSource) generateAndPruneIndexMergePath(indexMergeConds []expressio
+	for idx := 0; idx < normalPathCnt; idx++ {
+		if ds.possibleAccessPaths[idx].IsTablePath() || ds.possibleAccessPaths[idx].Index == nil || !ds.possibleAccessPaths[idx].Index.MVIndex {
+			continue // not a MVIndex path
+		}
+		if !ds.isSpecifiedInIndexMergeHints(ds.possibleAccessPaths[idx].Index.Name.L) {
+			continue // for safety, only consider using MVIndex when there is a `use_index_merge` hint now.
+			// TODO: remove this limitation
+		}
+
+		// Step 1. Extract the underlying JSON column from MVIndex Info.
+		mvIndex := ds.possibleAccessPaths[idx].Index
+		if len(mvIndex.Columns) != 1 {
+			// only support single-column MVIndex now: idx((cast(a->'$.zip' as signed array)))
+			// TODO: support composite MVIndex idx((x, cast(a->'$.zip' as int array), z))
+			continue
+		}
+		mvVirColOffset := mvIndex.Columns[0].Offset
+		mvVirColMeta := ds.table.Meta().Cols()[mvVirColOffset]
+
+		var virCol *expression.Column
+		for _, ce := range ds.TblCols {
+			if ce.ID == mvVirColMeta.ID {
+				virCol = ce.Clone().(*expression.Column)
+				virCol.RetType = ce.GetType().ArrayType() // use the underlying type directly: JSON-ARRAY(INT) --> INT
+				break
+			}
+		}
+		// unwrap the outside cast: cast(json_extract(test.t.a, $.zip), JSON) --> json_extract(test.t.a, $.zip)
+		targetJSONPath, ok := unwrapCast(virCol.VirtualExpr)
+		if !ok {
+			continue
+		}
+
+		// Step 2. Iterate all filters and generate corresponding IndexMerge paths.
+		for filterIdx, filter := range filters {
+			// Step 2.1. Extract jsonPath and vals from json_member / json_overlaps / json_contains functions.
+			sf, ok := filter.(*expression.ScalarFunction)
+			if !ok {
+				continue
+			}
+
+			var jsonPath expression.Expression
+			var vals []expression.Expression
+			switch sf.FuncName.L {
+			case ast.JSONMemberOf: // (1 member of a->'$.zip')
+				jsonPath = sf.GetArgs()[1]
+				v, ok := unwrapCast(sf.GetArgs()[0]) // cast(1 as json) --> 1
+				if !ok {
+					continue
+				}
+				vals = append(vals, v)
+			case ast.JSONOverlaps: // (json_overlaps(a->'$.zip', '[1, 2, 3]')
+				continue // TODO: support json_overlaps
+			case ast.JSONContains: // (json_contains(a->'$.zip', '[1, 2, 3]')
+				continue // TODO: support json_contains
+			default:
+				continue
+			}
+
+			// Step 2.2. Check some limitations.
+			if jsonPath == nil || len(vals) == 0 {
+				continue
+			}
+			if !jsonPath.Equal(ds.ctx, targetJSONPath) {
+				continue // not on the same JSON col
+			}
+			// only support INT now
+			// TODO: support more types
+			if jsonPath.GetType().EvalType() == types.ETInt {
+				continue
+			}
+			allInt := true
+			for _, v := range vals {
+				if v.GetType().EvalType() != types.ETInt {
+					allInt = false
+				}
+			}
+			if !allInt {
+				continue
+			}
+
+			// Step 2.3. Generate a IndexMerge Path of this filter on the current MVIndex.
+			var partialPaths []*util.AccessPath
+			for _, v := range vals {
```

Comment:
It looks like `IndexLookUp` is enough if there's only one value.

Reply:
Yes, but for simplicity we use IndexMerge to handle them(one-value or multiple-value) all.

Reply:
I leave a comment here and leave it as further optimization.